cpp/src/arrow/compute/exec/exec_node_runner_impl.cc

```diff
@@ -0,0 +1,149 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#include "arrow/compute/exec/exec_node_runner_impl.h"
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/record_batch.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/util/async_util.h"
+#include "arrow/util/cancel.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+namespace compute {
+
+class SimpleSyncRunnerImpl : public ExecNodeRunnerImpl {
+ public:
+  explicit SimpleSyncRunnerImpl(ExecContext* ctx) : ctx_(ctx) {}
+
+  Status SubmitTask(std::function<Status()> task) override {
+    if (finished_.is_finished()) {
+      return Status::OK();
+    }
+    auto status = task();
+    if (!status.ok()) {
+      return status;
+    }
+    if (input_counter_.Increment()) {
+      this->MarkFinished();
+    }
+    return Status::OK();
+  }
+
+  void MarkFinished(Status status = Status::OK()) override {
+    this->finished_.MarkFinished(status);
+  }
+
+  void InputFinished(int total_batches) override {
+    if (input_counter_.SetTotal(total_batches)) {
+      this->MarkFinished();
+    }
+  }
+
+  void StopProducing() override {
+    if (input_counter_.Cancel()) {
+      this->MarkFinished();
+    }
+  }
+
+ protected:
+  // Executor context
+  ExecContext* ctx_;
+
+  // Counter for the number of batches received
+  AtomicCounter input_counter_;
+};
+
+class SimpleParallelRunner : public SimpleSyncRunnerImpl {
+ public:
+  explicit SimpleParallelRunner(ExecContext* ctx) : SimpleSyncRunnerImpl(ctx) {
+    executor_ = ctx->executor();
+  }
+
+  Status SubmitTask(std::function<Status()> task) override {
+    if (finished_.is_finished()) {
+      return Status::OK();
+    }
+    auto status = task_group_.AddTask([this, task]() -> Result<Future<>> {
+      return this->executor_->Submit(this->stop_source_.token(),
+                                     [task]() { return task(); });
+    });
+    if (!status.ok()) {
+      return status;
+    }
+    if (input_counter_.Increment()) {
+      this->MarkFinished();
+    }
+    return Status::OK();
+  }
+
+  void MarkFinished(Status status = Status::OK()) override {
+    if (!status.ok()) {
+      this->StopProducing();
+      if (input_counter_.Cancel()) {
+        this->MarkFinished();
+      }
+      return;
+    }
+    task_group_.End().AddCallback([this](const Status& status) {
+      if (!status.ok()) {
+        this->finished_.MarkFinished();
+      } else if (!this->finished_.is_finished()) {
+        this->finished_.MarkFinished(status);
+      }
+    });
+  }
+
+  void StopProducing() override {
+    this->stop_source_.RequestStop();
+    if (input_counter_.Cancel()) {
+      this->MarkFinished();
+    }
+  }
```

Comment:
`RequestStop` won't take hold immediately, it simply tells the task group to stop running any additional tasks but it won't/can't interrupt any currently running tasks.  If we mark finished right away here then there is a risk that you discard the exec plan while tasks are still running.  I think, even in error scenarios, we must wait for the task group to finish before marking `finished_` complete.

Reply:
`MapNode::MarkFinished()` does not mark finished right away. First It check if it is running in async mode or not, if it is async (has an executor) then it adds a callback  to the `task_group`. Maybe the name is not clear for the porpuse of this logic